File: apps/api/app/routers/events.py
from uuid import UUID

from fastapi import APIRouter, Depends, Query, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.db.session import get_db
from app.models.user import User
from app.routers.auth import get_current_user
from app.schemas.events import (
    AttendanceEntry,
    AttendanceListResponse,
    CreateEventRequest,
    EventDetail,
    EventListResponse,
    HostAttendanceUpdateRequest,
    SelfAttendanceRequest,
)
from app.services import events as events_service

router = APIRouter(prefix="/events", tags=["events"])


@router.post("", response_model=EventDetail, status_code=status.HTTP_201_CREATED)
async def create_event(
    body: CreateEventRequest,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> EventDetail:
    return await events_service.create_event(db, current_user.id, body)


@router.get("", response_model=EventListResponse)
async def list_events(
    mine: bool = Query(False, description="Only events the caller hosts or has joined"),
    sport: str | None = Query(None),
    mode: str | None = Query(None, description="casual or ranked"),
    limit: int = Query(20, ge=1, le=50),
    offset: int = Query(0, ge=0),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> EventListResponse:
    return await events_service.list_events(
        db=db,
        current_user_id=current_user.id,
        mine=mine,
        sport=sport,
        mode=mode,
        limit=limit,
        offset=offset,
    )


@router.get("/{event_id}", response_model=EventDetail)
async def get_event(
    event_id: UUID,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> EventDetail:
    return await events_service.get_event(db, event_id, current_user.id)


@router.post("/{event_id}/join", response_model=EventDetail)
async def join_event(
    event_id: UUID,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> EventDetail:
    return await events_service.join_event(db, event_id, current_user.id)


@router.post("/{event_id}/leave", response_model=EventDetail)
async def leave_event(
    event_id: UUID,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> EventDetail:
    return await events_service.leave_event(db, event_id, current_user.id)


@router.post("/{event_id}/cancel", response_model=EventDetail)
async def cancel_event(
    event_id: UUID,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> EventDetail:
    return await events_service.cancel_event(db, event_id, current_user.id)


@router.post("/{event_id}/complete", response_model=EventDetail)
async def complete_event(
    event_id: UUID,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> EventDetail:
    return await events_service.complete_event(db, event_id, current_user.id)


@router.get("/{event_id}/attendance", response_model=AttendanceListResponse)
async def get_event_attendance(
    event_id: UUID,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> AttendanceListResponse:
    return await events_service.get_event_attendance(db, event_id, current_user.id)


@router.post("/{event_id}/attendance", response_model=AttendanceEntry)
async def host_update_attendance(
    event_id: UUID,
    body: HostAttendanceUpdateRequest,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> AttendanceEntry:
    return await events_service.host_update_attendance(
        db, event_id, current_user.id, body
    )


@router.post("/{event_id}/attendance/self", response_model=AttendanceEntry)
async def self_report_attendance(
    event_id: UUID,
    body: SelfAttendanceRequest,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> AttendanceEntry:
    return await events_service.self_report_attendance(
        db, event_id, current_user.id, body
    )

File: apps/mobile/src/lib/events.ts
/**
 * Events ("battles" in the UI) API client.
 *
 * Wraps the typed /events surface. The shared @protin/shared-types
 * package owns the response shapes — this file is intentionally thin
 * so the type contract stays in one place.
 */

import { api } from './api';
import type {
  AttendanceEntry,
  AttendanceListResponse,
  AttendanceStatus,
  CreateEventRequest,
  EventDetail,
  EventListResponse,
  EventMode,
  HostAttendanceUpdateRequest,
  SelfAttendanceRequest,
} from '@protin/shared-types';

export type {
  CreateEventRequest,
  EventDetail,
  EventListResponse,
  EventMode,
  EventStatus,
  EventSummary,
  EventVisibility,
  EventParticipantSummary,
  AttendanceStatus,
  SelfAttendanceStatus,
  ParticipantLifecycleStatus,
  AttendanceEntry,
  AttendanceListResponse,
  HostAttendanceUpdateRequest,
  SelfAttendanceRequest,
} from '@protin/shared-types';

export interface ListEventsParams {
  mine?: boolean;
  sport?: string;
  mode?: EventMode;
  limit?: number;
  offset?: number;
}

function buildQuery(params: ListEventsParams): string {
  const qs = new URLSearchParams();
  if (params.mine) qs.set('mine', 'true');
  if (params.sport) qs.set('sport', params.sport);
  if (params.mode) qs.set('mode', params.mode);
  if (params.limit !== undefined) qs.set('limit', String(params.limit));
  if (params.offset !== undefined) qs.set('offset', String(params.offset));
  const out = qs.toString();
  return out ? `?${out}` : '';
}

export async function listEvents(
  params: ListEventsParams = {}
): Promise<EventListResponse> {
  return api.get<EventListResponse>(`/events${buildQuery(params)}`);
}

export async function getEvent(eventId: string): Promise<EventDetail> {
  return api.get<EventDetail>(`/events/${eventId}`);
}

export async function createEvent(body: CreateEventRequest): Promise<EventDetail> {
  return api.post<EventDetail>('/events', body);
}

export async function joinEvent(eventId: string): Promise<EventDetail> {
  return api.post<EventDetail>(`/events/${eventId}/join`);
}

export async function leaveEvent(eventId: string): Promise<EventDetail> {
  return api.post<EventDetail>(`/events/${eventId}/leave`);
}

/**
 * Host-only: cancel the event. Backend returns the updated EventDetail
 * with status='cancelled'. Idempotent — calling on an already-cancelled
 * event returns the current detail without erroring.
 */
export async function cancelEvent(eventId: string): Promise<EventDetail> {
  return api.post<EventDetail>(`/events/${eventId}/cancel`);
}

/**
 * Host-only: complete the event. Backend returns status='completed'.
 * Rejected (422) if called before `starts_at`. Idempotent if already
 * completed.
 */
export async function completeEvent(eventId: string): Promise<EventDetail> {
  return api.post<EventDetail>(`/events/${eventId}/complete`);
}

/**
 * True if the event's `startsAt` is at or before now. Mirrors the
 * backend time-gate so the mobile UI hides attendance controls before
 * the game starts.
 */
export function eventHasStarted(startsAt: string): boolean {
  const ts = Date.parse(startsAt);
  if (Number.isNaN(ts)) return true;
  return ts <= Date.now();
}

// ---------------------------------------------------------------------------
// Attendance
// ---------------------------------------------------------------------------

export async function getEventAttendance(
  eventId: string
): Promise<AttendanceListResponse> {
  return api.get<AttendanceListResponse>(`/events/${eventId}/attendance`);
}

export async function hostUpdateAttendance(
  eventId: string,
  body: HostAttendanceUpdateRequest
): Promise<AttendanceEntry> {
  return api.post<AttendanceEntry>(`/events/${eventId}/attendance`, body);
}

export async function selfReportAttendance(
  eventId: string,
  body: SelfAttendanceRequest
): Promise<AttendanceEntry> {
  return api.post<AttendanceEntry>(`/events/${eventId}/attendance/self`, body);
}

export function attendanceStatusLabel(s: AttendanceStatus): string {
  switch (s) {
    case 'pending':
      return 'Pending';
    case 'attended':
      return 'Attended';
    case 'no_show':
      return 'No-show';
    case 'excused':
      return 'Excused';
  }
}

// ---------------------------------------------------------------------------
// Sport vocabulary
// ---------------------------------------------------------------------------

/**
 * Battle sport options. Keep in sync with the brief. Backend stores
 * sport as freeform lowercase string, so adding more here doesn't
 * require a migration.
 */
export const BATTLE_SPORTS = [
  { value: 'basketball', label: 'Basketball' },
  { value: 'soccer', label: 'Soccer' },
  { value: 'running', label: 'Run' },
  { value: 'golf', label: 'Golf' },
  { value: 'badminton', label: 'Badminton' },
  { value: 'tennis', label: 'Tennis' },
] as const;

export type BattleSportValue = (typeof BATTLE_SPORTS)[number]['value'];

/** Default capacities by sport — used as initial value in the host form. */
export const SPORT_CAPACITY_DEFAULTS: Record<string, number> = {
  basketball: 10,
  soccer: 10,
  running: 30,
  golf: 4,
  badminton: 4,
  tennis: 2,
};

export function sportLabelForBattle(sport: string): string {
  const found = BATTLE_SPORTS.find((s) => s.value === sport);
  return found ? found.label : sport.charAt(0).toUpperCase() + sport.slice(1);
}

/** Compact "Sat 17 May · 09:00" style. */
export function formatEventWhen(iso: string): string {
  const d = new Date(iso);
  const date = d.toLocaleDateString(undefined, {
    weekday: 'short',
    month: 'short',
    day: 'numeric',
  });
  const time = d.toLocaleTimeString(undefined, {
    hour: '2-digit',
    minute: '2-digit',
  });
  return `${date} · ${time}`;
}
